docs/adr/009_discovery.md

```diff
@@ -0,0 +1,81 @@
+# 009. LoRaWAN Network Server (LNS) discovery
+
+**Feature**: [#1480](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1480)
+**Authors**: Bastian Burger, Daniele Maggio, Maggie Salak
+**Status**: Accepted
+
+Service (or LNS) discovery is part of the LNS protocol. When a LoRa Basics Station (LBS) connects to the first time to a LNS, it invokes the service discovery endpoint (`/router-info`). We want to provide the users with the option to (potentially) increase availability by automatically rebalancing connection attempts to different LNS.
+
+We can consider implementing the service discovery in two places, because:
+
+- the LNS discovery endpoint is returned as a response as part of the CUPS protocol
+- the LNS data endpoint is returned as a response of the service discovery endpoint invocation (`/router-info`)
+
+we could consider implementing the service discovery as part of either the CUPS protocol or the LNS protocol. Since LBS does not reconnect to the CUPS endpoint when a connection to a LNS is dropped, we are only left with the option of implementing service discovery as part of the LNS service discovery endpoint.
+
+This implies that the discovery service needs to support WebSockets. As of now, the `/router-info` endpoint is part of the LNS itself. Since we want to use it to protect against LNS outages, by definition it becomes clear that discovery needs to be isolated from the LNS and needs to be come a standalone, highly available service. In the following, we propose several properties of the discovery service.
+
+## Decision
+
+We propose to add an ASP.NET Core Web Application to the OSS starter kit that exposes an endpoint for service discovery. The Web App can deployed anywhere (as a highly available cloud service per default, or as an on-premises service for more flexibility). With respect to configuration and health probe, we will implement two simple approaches initially, and expand the functionality in a second stage.
+
+In the initial version, we will not implement a health check. We will rely on the fact that if a LBS reconnects to the discovery service, the LNS was not available. By using a round-robin distribution mechanism based on in-memory state, we can guarantee reasonably well that the LBS will be connected to a different LNS in a second attempt.
+
+For configuring which LNS is in reach of a specific LBS, we are going to tag the LNS twins and station twins with a "location" or "network" tag. All twins with the same tag are considered to be in the same network and in reach of each other.
+
+In a second stage, we will prioritize one of the more advanced health probe strategies and potentially introduce more supported configuration approaches.
+
+We will maintain the discovery endpoints on the LNS for backwards compatibility.
```

Comment:
I think that should not be backwards compatibility but a real option. If you want to keep it low key and not deploy the app service, it's totally fine to keep the discovery service on the LNS.

Reply:
Totally agree, adapted in 5a2bd2762d2c4982a07d8e65a5c5d5f1e7de7702